### ISSUE TITLE
False positive in self_in_property_initialization

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#4041]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Linting Swift files at paths file.swift
Linting 'file.swift' (1/1)
/Users/marcelo.fabri/project/file.swift:2:5: warning: Self in Property Initialization Violation: `self` refers to the unapplied `NSObject.self()` method, which is likely not expected. Make the variable `lazy` to be able to refer to the current instance or use `ClassName.self`. (self_in_property_initialization)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.48.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? `portable_zip`
* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

No

* Which Xcode version are you using (check `xcodebuild -version`)? 

13.4.1

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
class Foo {
    var bar: Bool = false {
        didSet {
            value = {
                if bar {
                    return self.calculateA()
                } else {
                    return self.calculateB()
                }
            }()
            print(value)
        }
    }

    var value: String?

    func calculateA() -> String { "A" }
    func calculateB() -> String { "B" }
}
```
